nuxeo-features/nuxeo-platform-notification/nuxeo-platform-notification-core/src/main/java/org/nuxeo/ecm/platform/ec/notification/email/EmailHelper.java

```diff
@@ -185,38 +184,30 @@ protected void sendmail0(Map<String, Object> mail)
      * Gets the session from the JNDI.
      */
     private static Session getSession() {
-        Session session = null;
-        if (javaMailNotAvailable) {
-            return null;
-        }
-        // First, try to get the session from JNDI, as would be done under J2EE.
-        try {
-            NotificationService service = (NotificationService) Framework.getRuntime()
-                                                                         .getComponent(NotificationService.NAME);
-            InitialContext ic = new InitialContext();
-            session = (Session) ic.lookup(service.getMailSessionJndiName());
-        } catch (NamingException ex) {
-            log.warn("Unable to find Java mail API", ex);
-            javaMailNotAvailable = true;
+        if (!javaMailNotAvailable) {
+            // First, try to get the session from JNDI, as would be done under J2EE.
+            try {
+                NotificationService service = (NotificationService) Framework.getRuntime()
+                                                                             .getComponent(NotificationService.NAME);
+                return MailSessionBuilder.fromJndi(service.getMailSessionJndiName()).build();
+            } catch (NuxeoException ex) {
+                log.warn("Unable to find Java mail API", ex);
+                javaMailNotAvailable = true;
+            }
```

Comment:
i know that the var `javaMailNotAvailable` was here before you rework it and is was set to `true` if we don't find the `jndi` which means the next time is called the code returns `null` even the jndi is added programmatically -> what i mean is that `EmailHelper` is not reentrant. i know this an old code but i want to have you point of view ;)

Reply:
Indeed, what is important to me there is not to change behavior, issue you're talking about was there before. So not changing that is ok.

To go further, I'm not sure we have such case of adding new mail session to jndi context. To me, jndi resources are mainly initialised during web application container initialisation. Furthermore, if we want to programmatically build a mail session, why not using directly the builder instead of building one, binding it to jndi and then use this helper?

Reply:
yes  i agree that its now more cleaner and reworked i agree we can use the builders. and yes i agree that it's mainly initialised within the web application container :)